core/communication/port-authority/src/port.ts

```diff
@@ -43,8 +43,9 @@ export interface SendPort<Out> {
 /**
  * Map operation for [[SendPort]]s. The returned port behaves identically to the original port, but applies a function
  * to outgoing messages _before_ they are sent on the original port.
+ * @returns an instance of [[SendPort]] instantiated to the `In` class.
  */
-export function mapSendPort<Out, Out2>(port: SendPort<Out>, f: (x: Out2) => Out): SendPort<Out2> {
+export function mapSendPort<Out, In>(port: SendPort<Out>, f: (x: In) => Out): SendPort<In> {
```

Comment:
I'm going through this again and I think that Out and In are misleading. I think we should use a more generic type `T` or `Type`, or `T1, T2` in case there are two types involved.

Reply:
Well, I am referring to the fact that this map is  not **covariant**, but rather **contravariant**, as it is mapping on the Input, not the Output. I just mentioned this, if we are ok to use the same name for both "variants". I saw that some call it "contramap" in TypeScript.

Reply:
Well, depending on what you think it maps, which is part of the confusion. I think that the original intention was to map a port from a Port that takes Out to another that takes Out2. Right, confusing. But, at the same time it's going to be a (half) port that receives Out2 (right, confusing) and produces Out.
This is now called `txMappingPort<T, In>(port: TxPort<T>, f: (x: In) => T): TxPort<In>`. This port is a mapping port that receives `In` (sounds about right) and produces T (which is generic). Please check the new implementation.

Reply:
It is not about the implementation, it is simply that this map, maps on Output.

Consider this case of TxPort, where the output is also a value not just `void`:
```
export interface TxPort<A, B> {
     send(value: A): B;
 }
```
A covariant map, maps on the output B, B1, like this:
```
export function txMap<A, B, B1>(port: TxPort<A, B>, f: (x: B) => B1): TxPort<A, B1> {
     return {
         send: (value: A) => f(port.send(value))
     };
 }
```
A contravariant looks like this, which maps inputs A1, A

```
export function txContraMap<A, B, A1>(port: TxPort<A, B>, f: (x: A1) => A): TxPort<A1, B> {
     return {
         send: (value: A1) => post.send(f(value))
     };
 }
```

Reply:
Again this is not super important, just the same name "map" is used for both types. If it is not confusing to use "map" for both, then we should be ok. I am fairly comfortable with both types, I can spot the difference from the method signature and I know what to expect. But maybe having explicit name, would be better - just an idea.